File: XamarinStore.Droid/Views/ViewSwipeTouchListener.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Android.App;
using Android.Content;
using Android.OS;
using Android.Runtime;
using Android.Views;
using Android.Widget;
using Android.Animation;
using Android.Graphics;
using Android.Graphics.Drawables;

using Runnable = Java.Lang.Runnable;

namespace XamarinStore
{
	public class ViewSwipeTouchListener : GestureDetector.SimpleOnGestureListener, View.IOnTouchListener
	{
		GestureDetector detector;
		View targetView;
		ViewConfiguration config;
		int subviewID;

		public event EventHandler SwipeGestureBegin;
		public event EventHandler SwipeGestureEnd;
		public event EventHandler ItemSwipped;

		public ViewSwipeTouchListener (Context context, int subviewID)
		{
			this.detector = new GestureDetector (context, this);
			this.config = ViewConfiguration.Get (context);
			this.subviewID = subviewID;
		}

		public bool OnTouch (View v, MotionEvent e)
		{
			if (targetView == null)
				targetView = subviewID == 0 ? v : v.FindViewById (subviewID);
			if (e.Action == MotionEventActions.Up || e.Action == MotionEventActions.Cancel) {
				if (SwipeGestureEnd != null)
					SwipeGestureEnd (this, EventArgs.Empty);
				var dismiss = e.Action != MotionEventActions.Cancel
				              && targetView.TranslationX > targetView.Width / 2;
				SnapView (dismiss);
			}
			detector.OnTouchEvent (e);
			return true;
		}

		public void ResetSwipe ()
		{
			if (targetView != null) {
				targetView.Alpha = 1;
				targetView.TranslationX = 0;
			}
		}

		void SnapView (bool dismiss = true)
		{
			if (targetView == null)
				return;
			var targetAlpha = dismiss ? 0 : 1;
			var targetTranslation = dismiss ? targetView.Width : 0;
			var a = targetView.Animate ()
				.Alpha (targetAlpha)
				.TranslationX (targetTranslation);
			if (dismiss) {
				a.WithEndAction (new Runnable (() => {
					if (ItemSwipped != null)
						ItemSwipped (this, EventArgs.Empty);
				}));
			}
			a.Start ();
		}

		public override bool OnFling (MotionEvent e1, MotionEvent e2, float velocityX, float velocityY)
		{
			// We are only interested in an horizontal right-side fling
			if (velocityY > velocityX || velocityX < 0)
				return base.OnFling (e1, e2, velocityX, velocityY);

			SnapView (dismiss: true);
			return true;
		}

		public override bool OnScroll (MotionEvent e1, MotionEvent e2, float distanceX, float distanceY)
		{
			if (SwipeGestureBegin != null)
				SwipeGestureBegin (this, EventArgs.Empty);
			distanceX = -distanceX;
			if (Math.Abs (distanceY) > Math.Abs (distanceX) + config.ScaledTouchSlop
				|| (distanceX < 0 && targetView.TranslationX <= 0))
				return base.OnScroll (e1, e2, distanceX, distanceY);
			targetView.TranslationX = Math.Max (0, targetView.TranslationX + distanceX);
			targetView.Alpha = (targetView.Width - targetView.TranslationX) / ((float)targetView.Width);
			return true;
		}
	}
}

File: F#/XamarinStore.Droid/Views/ViewSwipeTouchListener.fs
namespace XamarinStore

open System
open Android.Content
open Android.Views
open Android.Animation

type ViewSwipeTouchListener(context:Context, subviewID:int) as this =
    inherit GestureDetector.SimpleOnGestureListener()

    let mutable detector: GestureDetector = new GestureDetector (context, this)
    let mutable targetView:View = null
    let mutable config:ViewConfiguration = ViewConfiguration.Get (context)
    let mutable subViewId = subviewID

    member this.ResetSwipe () =
        if not (targetView = null) then
            targetView.Alpha <- 1.0f
            targetView.TranslationX <- 0.0f


    member val SwipeGestureBegin = fun ()->() with get,set
    member val SwipeGestureEnd = fun ()->() with get,set
    member val ItemSwipped = fun ()->() with get,set

    member this.SnapView dismiss =
        if not (targetView = null) then
            let targetAlpha = if dismiss then 0.0f else 1.0f
            let targetTranslation = if dismiss then float32 targetView.Width else 0.0f
            let a = ObjectAnimator.OfPropertyValuesHolder (
                        targetView,
                        PropertyValuesHolder.OfFloat ("alpha", targetView.Alpha, targetAlpha),
                        PropertyValuesHolder.OfFloat ("translationX", targetView.TranslationX, targetTranslation))
            if dismiss then
                a.AnimationEnd.AddHandler(new EventHandler(fun sender eventargs ->(sender :?> ValueAnimator).RemoveAllListeners ()
                                                                                  this.ItemSwipped ()))
            a.Start () 

    interface View.IOnTouchListener with
        member this.OnTouch(v, e) =
            if (targetView = null) then
                targetView <- if subviewID = 0 then v else v.FindViewById subviewID
            if (e.Action = MotionEventActions.Up || e.Action = MotionEventActions.Cancel) then
                this.SwipeGestureEnd ()
                let dismiss = (not (e.Action = MotionEventActions.Cancel))
                              && (targetView.TranslationX > float32 targetView.Width / 2.0f)
                this.SnapView (dismiss)
            detector.OnTouchEvent (e) |> ignore
            true

    override this.OnFling(e1, e2, velocityX, velocityY) =
        // We are only interested in an horizontal right-side fling
        if (velocityY > velocityX || velocityX < 0.0f) then
            base.OnFling (e1, e2, velocityX, velocityY)
        else
            this.SnapView true
            true

    override this.OnScroll (e1, e2, distanceX, distanceY) =
        this.SwipeGestureBegin ()
        let distanceX = -distanceX
        if Math.Abs(distanceY) > (Math.Abs(distanceX) + float32 config.ScaledTouchSlop)
            || (distanceX < 0.0f && targetView.TranslationX <= 0.0f) then
            base.OnScroll (e1, e2, distanceX, distanceY)
        else targetView.TranslationX <- Math.Max (0.0f, targetView.TranslationX + distanceX)
             targetView.Alpha <- (float32 targetView.Width - targetView.TranslationX) / (float32 targetView.Width)
             true


